pubsub/pubsub.go

```diff
@@ -0,0 +1,208 @@
+	return func(s *Server) {
+		if cap > 0 {
+			s.cmds = make(chan cmd, cap)
+		}
+	}
+}
+
+// OverflowStrategyDrop will tell the server to drop messages when it can't
+// process more messages.
+func OverflowStrategyDrop() func(*Server) {
+	return func(s *Server) {
+		s.overflowStrategy = drop
+	}
+}
+
+// OverflowStrategyWait will tell the server to block and wait for some time
+// for server to process other messages. Default strategy.
+func OverflowStrategyWait() func(*Server) {
+	return func(s *Server) {
+		s.overflowStrategy = wait
+	}
+}
+
+// Subscribe returns a channel on which messages matching the given query can
+// be received.
+func (s *Server) Subscribe(query Query, out chan<- interface{}) {
+	s.cmds <- cmd{op: sub, ch: out, query: query}
+}
+
+// Unsubscribe unsubscribes the given channel.
+func (s *Server) Unsubscribe(out chan<- interface{}) {
+	s.cmds <- cmd{op: unsub, ch: out}
+}
+
+// Publish publishes the given message.
+func (s *Server) Publish(msg interface{}) error {
+	return s.PublishWithTags(msg, make(map[string]interface{}))
+}
+
+// PublishWithTags publishes the given message with a set of tags. This set of
+// tags will be matched with client queries. If there is a match, the message
+// will be sent to a client.
+func (s *Server) PublishWithTags(msg interface{}, tags map[string]interface{}) error {
+	pubCmd := cmd{op: pub, msg: msg, tags: tags}
+	if s.overflowStrategy == drop {
+		select {
+		case s.cmds <- pubCmd:
+		default:
+			s.Logger.Error("Failed to publish msg due to the server overflow")
+			return ErrorOverflow
+		}
+	} else if s.overflowStrategy == wait {
+		s.cmds <- pubCmd
+	}
+	return nil
+}
+
+// OnStop implements Service.OnStop by shutting down the server.
+func (s *Server) OnStop() {
+	s.cmds <- cmd{op: shutdown}
+}
+
+// NOTE: not goroutine safe
+type state struct {
+	queries map[chan<- interface{}]Query
+}
+
+// OnStart implements Service.OnStart by creating a main loop.
+func (s *Server) OnStart() error {
+	go s.loop(state{
+		queries: make(map[chan<- interface{}]Query),
+	})
+	return nil
+}
+
+func (s *Server) loop(state state) {
+loop:
+	for cmd := range s.cmds {
+		switch cmd.op {
+		case unsub:
+			state.remove(cmd.ch)
+		case shutdown:
+			state.removeAll()
+			break loop
+		case sub:
+			state.add(cmd.query, cmd.ch)
+		case pub:
+			state.send(cmd.msg, cmd.tags)
+		}
+	}
+}
+
+func (state *state) add(q Query, ch chan<- interface{}) {
+	state.queries[ch] = q
+}
+
+func (state *state) send(msg interface{}, tags map[string]interface{}) {
+	for ch, q := range state.queries {
+		// NOTE we can use LRU cache to speed up common cases like query = "
+		// tm.events.type=NewBlock" and tags = {"tm.events.type": "NewBlock"}
+		//
+		// Also, to reduce memory consumption we can store query string <-> Query
+		// pairs in an LRU cache (using reference counter to count number of
+		// clients using this query).
+		if q.Matches(tags) {
+			// non-blocking write
+			select {
+			case ch <- msg:
+			default:
```

Comment:
if the subscriber died/stopped OR is not able to keep up with the server, do we want to wait for it in any of these cases? maybe... the same set of strategies will do (wait, wait 10ms, drop)